### PR TITLE
Row group index caching

### PR DESCRIPTION
**Pull request by rongrong**
Test plan - TestOrcReader

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/test/java/com/facebook/presto/orc/AbstractTestOrcReader.java [hunk 1]
```diff
@@ -211,9 +213,15 @@ public void testCaching()
...
+        Cache<StripeReader.StripeStreamId, List<RowGroupIndex>> rowGroupIndexCache = CacheBuilder.newBuilder()
+                .maximumWeight(new DataSize(1, MEGABYTE).toBytes())
+                .weigher((id, rowGroupIndices) -> toIntExact(((List<RowGroupIndex>) rowGroupIndices).stream().mapToLong(RowGroupIndex::getRetainedSizeInBytes).sum()))
+                .expireAfterAccess(new Duration(10, MINUTES).toMillis(), MILLISECONDS)
+                .recordStats()
+                .build();
+        StripeMetadataSource stripeMetadataSource = new CachingStripeMetadataSource(new StorageStripeMetadataSource(), stripeFootercache, stripeStreamCache, rowGroupIndexCache);
 
-        try (TempFile tempFile = createTempFile()) {
+        try (TempFile tempFile = createTempFile(20000)) {
```
Comment (yuanzhanhku) [inline presto-orc/src/test/java/com/facebook/presto/orc/AbstractTestOrcReader.java:224]: Does this actual number of records matter in the test? Could we reduce it? I am wondering if writing 20000 record is too expensive in a unit test.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 2]
```diff
@@ -299,11 +301,19 @@ public StripeMetadataSourceFactory createStripeMetadataSourceFactory(OrcCacheCon
                     .expireAfterAccess(orcCacheConfig.getStripeStreamCacheTtlSinceLastAccess().toMillis(), MILLISECONDS)
                     .recordStats()
                     .build();
+            Cache<StripeStreamId, List<RowGroupIndex>> rowGroupIndexCache = CacheBuilder.newBuilder()
```
Comment (yuanzhanhku) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:304]: orc.stripe-metadata-cache-enabled is enabled by default in some raptorX clusters. But we may not want to enable the new RowGroupIndex by default without testing it properly. Maybe add a new session property to control it?

File: presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java [hunk 3]
```diff
@@ -514,26 +511,20 @@ static boolean isIndexStream(Stream stream)
...
-                        RowGroupIndex rowGroupIndex = rowGroupIndexes.get(i);
-                        ColumnStatistics columnStatistics = rowGroupIndex.getColumnStatistics()
-                                .withBloomFilter(bloomFilters.get(i));
-                        newRowGroupIndexes.add(new RowGroupIndex(rowGroupIndex.getPositions(), columnStatistics));
-                    }
-                    rowGroupIndexes = newRowGroupIndexes.build();
-                }
+                OrcInputStream inputStream = streamsData.get(streamId);
+                List<HiveBloomFilter> bloomFilters = bloomFilterIndexes.get(streamId.getColumn());
+                List<RowGroupIndex> rowGroupIndexes = stripeMetadataSource.getRowIndexes(metadataReader, hiveWriterVersion, stripeId, streamId, inputStream, bloomFilters);
```
Comment (yuanzhanhku) [inline presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java:527]: Maybe better to add some per query counters to get some insights on the size of the row index and cache hit rate when the cache is enabled.

StripeReader has a runtimeStats which can be passed into the getRowIndexes function and we could only populate the metrics in CachingStripeMetadataSource when the cache is enabled.

File: presto-orc/src/test/java/com/facebook/presto/orc/AbstractTestOrcReader.java (hunk 1 above)
Comment (rongrong) [reply presto-orc/src/test/java/com/facebook/presto/orc/AbstractTestOrcReader.java:224]: changed to 10001. The code path is only triggered when a stripe has more than 1 row group, and each row group has 10K rows. I didn't find a easy way to change the size of the row group.

**Review by highker — COMMENTED** [review]
Will leave it to @beinan, @yuanzhanhku, and @arunthirupathi to review and merge

**Review by yuanzhanhku — APPROVED** [review]
Looks good to me. Thanks for implementing this. 

Please add some detail the PR and the commit description. Mention that the new cache is disabled by default and list the config options which need to be set to enable/tune it.

**Review by beinan — APPROVED** [review]
lgtm, thank you @rongrong for also making it in iceberg connector!

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingStripeMetadataSource.java [hunk 4]
```diff
@@ -106,6 +116,36 @@ public Slice getStripeFooterSlice(OrcDataSource orcDataSource, StripeId stripeId
         return inputsBuilder.build();
     }
 
+    @Override
+    public List<RowGroupIndex> getRowIndexes(
+            MetadataReader metadataReader,
+            HiveWriterVersion hiveWriterVersion,
+            StripeId stripId,
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingStripeMetadataSource.java:123]: nit: stripeId

File: presto-orc/src/main/java/com/facebook/presto/orc/CachingStripeMetadataSource.java [hunk 5]
```diff
@@ -106,6 +116,36 @@ public Slice getStripeFooterSlice(OrcDataSource orcDataSource, StripeId stripeId
...
+        if (rowGroupIndexCache.isPresent()) {
+            List<RowGroupIndex> rowGroupIndices = rowGroupIndexCache.get().getIfPresent(new StripeStreamId(stripId, streamId));
+            if (rowGroupIndices != null) {
+                runtimeStats.addMetricValue("OrcRowGroupIndexCacheHit", 1);
+                runtimeStats.addMetricValue("OrcRowGroupIndexInMemoryBytesRead", rowGroupIndices.stream().mapToLong(RowGroupIndex::getRetainedSizeInBytes).sum());
+                return rowGroupIndices;
+            }
+            else {
+                runtimeStats.addMetricValue("OrcRowGroupIndexCacheHit", 0);
+                runtimeStats.addMetricValue("OrcRowGroupIndexStorageBytesRead", inputStream.getRetainedSizeInBytes());
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/CachingStripeMetadataSource.java:139]: This is incorrect metric.  OrcInputStream can be lazy and all data may not be loaded. It is better to add this metric after the rwGroupwIndices are decoded.

**Review by arunthirupathi — COMMENTED** [review]
Sorry, I was out of office last week. Thanks for contributing this feature.

I left couple of nits.

Do you know how the cache works for different files ? Different files will have same stripe Id (0,1,2) and StreamId (ROW_INDEX). Where is the file path used in the cache lookup ?

**Comment by yuanzhanhku** [comment]
> Do you know how the cache works for different files ? Different files will have same stripe Id (0,1,2) and StreamId (ROW_INDEX). Where is the file path used in the cache lookup ?

The StripeId has a OrcDataSourceId field which has the full file path, right?

**Comment by arunthirupathi** [comment]
> > Do you know how the cache works for different files ? Different files will have same stripe Id (0,1,2) and StreamId (ROW_INDEX). Where is the file path used in the cache lookup ?
> 
> The StripeId has a OrcDataSourceId field which has the full file path, right?

Nice, this makes sense now.

**Pull request merged**